x-pack/plugins/alerting_builtins/server/alert_types/geo_threshold/geo_threshold.ts

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import _ from 'lodash';
+import { SearchResponse } from 'elasticsearch';
+import { executeEsQueryFactory, getShapesFilters, OTHER_CATEGORY } from './es_query_builder';
+import { AlertServices, AlertTypeState } from '../../../../alerts/server';
+import { ActionGroupId } from './alert_type';
+import { Logger } from '../../types';
+
+interface LatestEntityLocation {
+  location: string;
+  shapeLocationId: string;
+  entityName: string;
+  dateInShape: string;
+  docId: string;
+}
+
+// Flatten agg results and get latest locations for each entity
+function transformResults(
+  results: SearchResponse<unknown> | undefined,
+  dateField: string,
+  geoField: string
+): LatestEntityLocation[] {
+  if (!results) {
+    return [];
+  }
+
+  return (
+    _.chain(results)
+      .get('aggregations.shapes.buckets', {})
+      // @ts-expect-error
+      .flatMap((bucket: unknown, bucketKey: string) => {
+        const subBuckets = _.get(bucket, 'entitySplit.buckets', []);
+        return _.map(subBuckets, (subBucket) => ({
+          location: _.get(subBucket, `entityHits.hits.hits[0].fields.${geoField}`, null),
+          shapeLocationId: bucketKey,
+          entityName: subBucket.key,
+          dateInShape: _.get(subBucket, `entityHits.hits.hits[0].fields.${dateField}[0]`, null),
+          docId: _.get(subBucket, `entityHits.hits.hits[0]._id`),
+        }));
+      })
+      .orderBy(['entityName', 'dateInShape'], ['asc', 'desc'])
+      .sortedUniqBy('entityName')
+      .value()
+  );
+}
+
+interface EntityMovementDescriptor {
+  entityName: string;
+  currLocation: {
+    location: string;
+    shapeId: string;
+    date: string;
+    docId: string;
+  };
+  prevLocation: {
+    location: string;
+    shapeId: string;
+    date: string;
+    docId: string;
+  };
+}
+
+function getMovedEntities(
```

Comment:
consider exporting `getMovedEntities` and `transformResults` so these can be unit tested in isolation.